### ISSUE TITLE
Issues under high load & i/o pressure

### ISSUE THREAD

**Issue by ReToCode** [sematext/sematext-agent-docker#38]
Hi there,

We recently had a big issue with the sematext docker agent on our OpenShift platform. This issue might be tricky - I was not yet able to reproduce it on purpose. 

The issue occurs on our platform with 220 containers on one host system when a lot of those containers get high load and high log output at the same time. Those containers are not even logging to sematext, but the agent still is causing the problem. We were able to "fix" the problem with SKIP_BY_NAME=.*zvs.*|.*nova.* (the problematic containers).

So I checked the source code. It seems like SKIP_BY_NAME totally ignores those containers. Otherwise those containers are still attached and thus streamed, even if not necessary:

### Flow with SKIP_BY_NAME:
dockerLogsense.js > docker-loghose > docker-allcontainers: Then the containers are skipped directly & never attached & streamed

### Flow without SKIP_BY_NAME:
dockerLogsense.js > docker-loghose > docker-allcontainers:
- now all containers are returned to docker-loghose
- for every container now we do "docker inspect" & "docker attach" (at this point logs are already streamed!!)
- dockerLogsense listens for 'attach' events on docker and then checks if LOGSENE_ENABLED is enabled
- if not it waits for the first log line and then detaches the containers

As said above, I am not yet able to reproduce the issue on purpose, but skipping those containers "fixes" is. The flow without SKIP_BY_NAME seems not ideal to me, as every container is attached & streamed even if he is not enabled for logsene. Would it be possible to find another way of doing this? Maybe just return all container ids, then inspect them in logsense and only attach to them if necessary?

#### Btw, or logagent has those settings:
LOGSENE_ENABLED_DEFAULT=false
LOGSENE_ENABLED=true

Thank you in advance,
Regars Reto

@szakasz: FYI.

**Comment by megastef** [comment]
Thank you for reporting this issue. We made a PR to docker-loghose https://github.com/mcollina/docker-loghose/pull/25 to have an option to change the flow. This would help to avoid duplicated docker-inspect calls and log streams would be attached once we have implemented the filter function with LOGSENE_ENABLED logic.

**Comment by megastef** [comment]
@ReToCode SDA should not attach anymore to containers having LOGSENE_ENABLED=false - we use the new docker-loghose attachFilter function. This means we save the attach, detach API call and we save for each container 1 docker-inspect call because we share the result of docker-loghose inspect API call - for 220 containers this means we save up to 660 docker API calls.  We would appreciate if you could test `sematext/sematext-agent-docker:dev` with your workload.

**Comment by ReToCode** [comment]
Thanks we'll test this. Also if I can reproduce the "stuck" issue with a lot of attached containers, I'll get back to you.

**Comment by ReToCode** [comment]
I rolled out the "dev" tags on some node today. But some of them crashed pretty fast. Seems like there is an error somewhere in that build:

```
Container da0494ffc09b0a7590e1a0488ccf8128dbdad610139f1b9a5e560a33cb8560ad /k8s_POD.6960dc2b_touri-zug-keeper-72-9qwnk_touri-dev_265c5f9d-c63b-11e7-8a8f-1c98ec3025f0_ca6ea8a0 setting LOGSENE_ENABLED=false
--
  | SDA log stats: count=1997 shipped=756 bulks_req_send=9 bulk_req_failed=0 bulk_req_retransmit=0
  | SDA log stats: count=2397 shipped=2450 bulks_req_send=14 bulk_req_failed=0 bulk_req_retransmit=0
  | 2017-11-15T07:18:32.065Z - ERROR - Please contact support@sematext.com to report the error:
  | 2017-11-15T07:18:32.066Z - ERROR - UncaughtException:TypeError: dockerInspectHelper.getLogseneTokenByContainer is not a function
  | TypeError: dockerInspectHelper.getLogseneTokenByContainer is not a function
  | at DockerEvents.<anonymous> (/usr/src/app/lib/dockerLogsene.js:217:27)
  | at emitOne (events.js:116:13)
  | at DockerEvents.emit (events.js:211:7)
  | at JSuck.<anonymous> (/usr/src/app/node_modules/docker-events/index.js:34:12)
  | at emitOne (events.js:116:13)
  | at JSuck.emit (events.js:211:7)
  | at addChunk (_stream_readable.js:263:12)
  | at readableAddChunk (_stream_readable.js:250:11)
  | at JSuck.Readable.push (_stream_readable.js:208:10)
  | at JSuck.Transform.push (_stream_transform.js:147:32)
  | 2017-11-15T07:18:34.125Z - ERROR - Please contact support@sematext.com to report the error:
  | 2017-11-15T07:18:34.126Z - ERROR - UncaughtException:Error: Parse Error
  | Error: Parse Error
  | at Socket.socketOnData (_http_client.js:440:20)
  | at emitOne (events.js:116:13)
  | at Socket.emit (events.js:211:7)
  | at addChunk (_stream_readable.js:263:12)
  | at readableAddChunk (_stream_readable.js:250:11)
  | at Socket.Readable.push (_stream_readable.js:208:10)
  | at Pipe.onread (net.js:594:20)


```

**Comment by megastef** [comment]
@ReToCode Sorry, the wrong function call got fixed in `sematext/sematext-agent-docker:dev` image.  Please pull it again.

**Comment by ReToCode** [comment]
Seems to work now. So far no problems. We also further analyzed the root cause for the issue. We were able to reproduce the stuck log stream also on docker layer with 'docker logs -f'. If we put in enough log rows, the stream stops reporting data at some point, but it does not close or error out. We reported that issue to Redhat.

**Testcase:**
https://github.com/ReToCode/lb-test-backend

Then hit it with apache bench:
ab -n 10000 -c 10 http://url-to-container
